clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -1871,4 +1873,100 @@ void handleFailure(Throwable throwable) {
         return new CreatePartitionsResult(new HashMap<String, KafkaFuture<Void>>(futures));
     }
 
+    public DeleteRecordsResult deleteRecords(final Map<TopicPartition, RecordsToDelete> partitionsAndOffsets,
+                                             final DeleteRecordsOptions options) {
+
+        // requests need to be sent to partitions leader nodes so ...
+        // ... from the provided map it's needed to create more maps grouping topic/partition per leader
+
+        final Map<TopicPartition, KafkaFutureImpl<Long>> futures = new HashMap<>(partitionsAndOffsets.size());
+        for (TopicPartition topicPartition: partitionsAndOffsets.keySet()) {
+            futures.put(topicPartition, new KafkaFutureImpl<Long>());
+        }
+
+        // preparing topics list for asking metadata about them
+        final Set<String> topics = new HashSet<>();
+        for (TopicPartition topicPartition: partitionsAndOffsets.keySet()) {
+            topics.add(topicPartition.topic());
+        }
+
+        final long nowMetadata = time.milliseconds();
+        final long deadline = calcDeadlineMs(nowMetadata, options.timeoutMs());
+        // asking for topics metadata for getting partitions leaders
+        runnable.call(new Call("topicsMetadata", deadline,
+                new LeastLoadedNodeProvider()) {
+
+            @Override
+            AbstractRequest.Builder createRequest(int timeoutMs) {
+                return new MetadataRequest.Builder(new ArrayList<>(topics), false);
```

Comment:
I don't think this will work, will it?  You should specify the topics that you want metadata about.  If you specify an empty list, no topic info will be returned.

Reply:
With an empty list it means "return me every topic you know".

Reply:
null means "return me every topic you know".  The empty list means no topics.  (This changed in a previous AK version)

Reply:
> I don't think this will work, will it? You should specify the topics that you want metadata about. If you specify an empty list, no topic info will be returned.

@cmccabe what do you mean ? the `topics` list is built few lines before starting from the map provided. I'm pretty sure you noticed that so I'm missing your point here, sorry :-(

Reply:
Sorry, my earlier comment was related to an earlier version of the patch.